clients/drcachesim/tracer/instr_counter.cpp

```diff
@@ -347,9 +398,122 @@ event_inscount_thread_init(void *drcontext)
         DELAY_COUNTDOWN_UNIT;
 }
 
+// Represents an interval in terms of number of instructions.
+struct instr_interval_t {
+    instr_interval_t(uint64 start, uint64 duration)
+        : start(start)
+        , duration(duration)
+    {
+    }
+
+    uint64 start;
+    uint64 duration;
+};
+
+// Function to order instruction intervals by start time in ascending order.
+static bool
+cmp_start_instr(const instr_interval_t &l, const instr_interval_t &r)
+{
+    return l.start < r.start;
+}
+
+static std::vector<instr_interval_t>
+parse_instr_intervals_file(std::string path_to_file)
+{
+    std::vector<instr_interval_t> instr_intervals;
+    std::ifstream file(path_to_file);
```

Comment:
I think we want to avoid using these external libraries (std::vector and std::ifstream) in the online tracer code to avoid transparency issues caused by global state conflicts in these libraries (https://dynamorio.org/using.html#sec_extlibs). Check out the drcontainers, particularly drvector (https://dynamorio.org/page_drcontainers.html), and dr_read_file.

Reply:
Is this init-time-only?  Libraries are considered safe at init time (static library use is assumed to use app-triggered attach from a safe point).  See #2855 which added a test that fails if heap allocation via 3rd party library is added post-init.  You would think I would have documented this too but I'm not finding anything at the moment.

Also, "online tracer" sounds like as opposed to offline: the issue is with statically linking DR and the tracer into the app as then there are no shared library boundaries to use for library isolation control points.

Reply:
> static library use is assumed to use app-triggered attach from a safe point

The app-triggered attach point may be safe for the one thread that calls the dr_app_start etc APIs, but I don't think there's any guarantee that other threads are not in an unsafe spot (e.g., in the middle of a library routine).

> Also, "online tracer" sounds like as opposed to offline

I meant the online tracing work as opposed to offline post-processing (by raw2trace or analyzers) done by drmemtrace, which includes all modes of drmemtrace operation (like the other interpretation of online vs offline you pointed out).

Reply:
> > static library use is assumed to use app-triggered attach from a safe point
> 
> The app-triggered attach point may be safe for the one thread that calls the dr_app_start etc APIs, but I don't think there's any guarantee that other threads are not in an unsafe spot (e.g., in the middle of a library routine).

Right, but the process-wide init code is all run in that one thread, where it's considered safe.

Reply:
Makes sense.

Reply:
> You would think I would have documented this too but I'm not finding anything at the moment.

I agree with documenting this guidance with the rationale. I worry that the init time distinction may open up possible misuses of external libraries (from simple copy-pasting existing library use without knowledge of this limitation, or from not being properly informed what init-time means). But I guess as long as we document it clearly.

Reply:
>  I worry that the init time distinction may open up possible misuses of external libraries

That's what the regression test catches as nearly all the problems come from malloc.

Reply:
That test is limited to drmemtrace though. My concern was in other client tools not owned by us.

Reply:
Now using drvector_t for the `trace` and `no_trace` globals, since those are not just used during initialization.
Still using STL in init function for convenience, since it's allowed.
Where would it be a good place to document that libraries can be used in init routines?

Reply:
https://github.com/DynamoRIO/dynamorio/pull/6943

Reply:
Thank you, Derek.